### PR TITLE
Preorder traversal added

### PR DESCRIPTION
**Pull request by shivammahajancse**
### Have you read the [Contributing Guidelines on Pull Requests](https://github.com/TesseractCoding/NeoAlgo/blob/master/CONTRIBUTING.md#pull-requests)?

(Yes)

### Checklist

- [x] I've read the contribution guidelines.
- [x] I've checked the issue list before deciding what to submit.
- [x] I've edited the `README.md` and link to my code.

## Related Issues or Pull Requests

(#178 )

### CODE REVIEW

**Comment by shivammahajancse** [comment]
Please review sir @HarshCasper @atarax665 @ankitaggarwal23 @gat786

**Comment by shivammahajancse** [comment]
Sir @iamrajiv please review

**Comment by shivammahajancse** [comment]
Please review @nidh-ish

File: docs/data-structures/preordertraversal.mdx [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+title: Preorder traversal of Binary Tree
+slug: /data-structures/preordertraversal
+---
+
+## Introduction
+
+Binary Tree is a type of tree which have atmost two children or we can say atmost degree 2. So to traverse the binary tree according to Depth first search. We have a traversal algorithm called Preorder traversal.
+
+## Explanation 
+In this type of traversal(Preorder traversal). First we have to visit the root then its left part then right part respectively. In this first visit root then its left part that means we have to traverse whole left subtree after that whole right subtree. Preorder: (root->left->right).
```
Comment (iamrajiv) [inline docs/data-structures/preordertraversal.mdx:11]: @shivammahajancse Break a line after `## Explanation`.

File: docs/data-structures/preordertraversal.mdx [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+
+## Explanation 
+In this type of traversal(Preorder traversal). First we have to visit the root then its left part then right part respectively. In this first visit root then its left part that means we have to traverse whole left subtree after that whole right subtree. Preorder: (root->left->right). 
+### Example 1
+
+<img src="https://i.imgur.com/f0nHPpz.jpg?1" width="600" height="350"/>
+
+```
+
+ First traverse the root: Node(44).
```
Comment (iamrajiv) [inline docs/data-structures/preordertraversal.mdx:18]: Remove `.` after `Node(44)`.

File: docs/data-structures/preordertraversal.mdx [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+ Visit its root: Node(11)
+ Preorder traversal: 44 22 11
+ Visit right child of node(22): Node(33)
+ Preorder traversal: 44 22 11 33
+ Visit right part of node(44): Node(66)
+ Preorder traversal: 44 22 11 33 66 
+ Left child of this root: 55
+ Preorder traversal: 44 22 11 33 66 55
+ Right child of this subtree: 77
+ Now our Preorder traversal: 44 22 11 33 66 55 77 !!
```
Comment (iamrajiv) [inline docs/data-structures/preordertraversal.mdx:30]: Remove `!!` after `77`.

**Review by iamrajiv — CHANGES_REQUESTED** [review]
@shivammahajancse Rename the file to `preorder_traversal.mdx`.

File: docs/data-structures/preordertraversal.mdx [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+slug: /data-structures/preordertraversal
+---
+
+## Introduction
+
+Binary Tree is a type of tree which have atmost two children or we can say atmost degree 2. So to traverse the binary tree according to Depth first search. We have a traversal algorithm called Preorder traversal.
+
+## Explanation 
+In this type of traversal(Preorder traversal). First we have to visit the root then its left part then right part respectively. In this first visit root then its left part that means we have to traverse whole left subtree after that whole right subtree. Preorder: (root->left->right). 
+### Example 1
```
Comment (iamrajiv) [inline docs/data-structures/preordertraversal.mdx:12]: @shivammahajancse Below paragraphs look more grammatically correct please edit to this

```markdown
## Introduction

A binary tree is a type of tree which have at most two children or we can say almost degree 2. So to traverse the binary tree according to Depth-first search. We have a traversal algorithm called Preorder traversal.

## Explanation

In this type of traversal(Preorder traversal). First, we have to visit the root then the left part then the right part respectively. In this first visit root then its left part that means we have to traverse the whole left subtree after that whole right subtree. Preorder: (root->left->right).
```

**Review by iamrajiv — CHANGES_REQUESTED** [review]

**Comment by shivammahajancse** [comment]
All changes done sir @iamrajiv

**Review by iamrajiv — APPROVED** [review]

File: docs/data-structures/preorder_traversal.mdx [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+---
+
+## Introduction
+
+A binary tree is a type of tree which have at most two children or we can say at most degree 2. So to traverse the binary tree according to Depth-first search. We have a traversal algorithm called Preorder traversal
+
+## Explanation 
+
+In this type of traversal(Preorder traversal). First, we have to visit the root then the left part then the right part respectively. In this first visit root then its left part that means we have to traverse the whole left subtree after that whole right subtree. Preorder: (root->left->right).
+### Example 1
```
Comment (nidh-ish) [inline docs/data-structures/preorder_traversal.mdx:13]: Use heading 4.

File: docs/data-structures/preorder_traversal.mdx [hunk 6]
```diff
@@ -0,0 +1,210 @@
+---
+title: Preorder traversal of Binary Tree
+slug: /data-structures/preorder_traversal
+---
+
+## Introduction
+
+A binary tree is a type of tree which have at most two children or we can say at most degree 2. So to traverse the binary tree according to Depth-first search. We have a traversal algorithm called Preorder traversal
```
Comment (nidh-ish) [inline docs/data-structures/preorder_traversal.mdx:8]: Add a full stop at the end.

File: docs/data-structures/preorder_traversal.mdx [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+ Preorder traversal: 44 22 11 33
+ Visit right part of node(44): Node(66)
+ Preorder traversal: 44 22 11 33 66 
+ Left child of this root: 55
+ Preorder traversal: 44 22 11 33 66 55
+ Right child of this subtree: 77
+ Now our Preorder traversal: 44 22 11 33 66 55 77 
+
+```
+# Algorithm
```
Comment (nidh-ish) [inline docs/data-structures/preorder_traversal.mdx:34]: Use heading 2.

File: docs/data-structures/preorder_traversal.mdx [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+
+* First we have to traverse the Binary tree in this order(root->left->right).
+
+* Left means traverse the whole left subtree in this order(root->left->right) until we get left child as null. 
+
+* Root means parent of that particular subtree.
+
+* Right means traverse the whole right subtree in this (root->left->right) order until we get null.
+
+# Code
```
Comment (nidh-ish) [inline docs/data-structures/preorder_traversal.mdx:46]: Use heading 2.

File: docs/data-structures/preorder_traversal.mdx [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+
+# Preorder Traversal :- left -> root -> right
+
+Preorder(tree)			
+    
+```
+</TabItem>
+</Tabs>
+
+# Sample Input and Output
```
Comment (nidh-ish) [inline docs/data-structures/preorder_traversal.mdx:186]: Use heading 2.

File: docs/data-structures/preorder_traversal.mdx [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+Enter the data: 44 22 66 11 33 55 77 
+
+```
+#### Output :
+
+```
+Preorder traversal : 44 22 11 33 66 55 77
+```
+
+# Time & Space Complexity
```
Comment (nidh-ish) [inline docs/data-structures/preorder_traversal.mdx:201]: Make it: Complexity Analysis, and use heading 2

File: docs/data-structures/preorder_traversal.mdx [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+ 
+- **Time Complexity:** `O(n)`
+
+- **Space Complexity:** `O(h)`
+
+where h is the height of binary tree and n is the number of elements.
```
Comment (nidh-ish) [inline docs/data-structures/preorder_traversal.mdx:207]: Keep h and n in backticks.

File: docs/data-structures/preorder_traversal.mdx [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+
+For Preorder traversal, we follow the following algorithm:
+
+* First we have to traverse the Binary tree in this order(root->left->right).
+
+* Left means traverse the whole left subtree in this order(root->left->right) until we get left child as null. 
+
+* Root means parent of that particular subtree.
+
+* Right means traverse the whole right subtree in this (root->left->right) order until we get null.
```
Comment (nidh-ish) [inline docs/data-structures/preorder_traversal.mdx:44]: Make it:

For Preorder traversal, we follow the following algorithm:

First, we have to traverse the Binary tree in this order(root->left->right).

Left means traverse the whole left subtree in this order(root->left->right) until we get the left child as null.

Root means the parent of that particular subtree.

Right means traverse the whole right subtree in this (root->left->right) order until we get null.

File: docs/data-structures/preorder_traversal.mdx (hunk 8 above)
Comment (nidh-ish) [inline docs/data-structures/preorder_traversal.mdx:46]: Make the input/output uniform across both the codes as you did in the in-order traversal.

**Review by nidh-ish — CHANGES_REQUESTED** [review]

**Comment by shivammahajancse** [comment]
All changes done @nidh-ish

**Review by nidh-ish — APPROVED** [review]
LGTM

**Review by iamrajiv — APPROVED** [review]

**Pull request merged**